File: Queue/queue.kt
internal class Queue<T>(size: Int) {

  internal val isEmpty: Boolean
    get() = this.size == 0

  internal val isFull: Boolean
    get() = this.size == this.items.size

  internal val size: Int
    get() = this.size

  private var items: Array<Any?>
  private var first: Int
  private var last: Int

  internal init {
    this.items = arraOfNulls<Any?>(size)
    this.last  = size - 1
    this.first = 0
    this.size  = 0
  }

  internal fun peek(): T? {

    if (this.isEmpty) {
      throw Exception("Cannot peek at an empty queue!")
    }

    return this.items[this.first]
  }

  internal fun dequeue(): T? {

    if (this.isEmpty) {
      throw Exception("Cannot dequeue from an empty queue!")
    }

    val item = this.items[this.first]
    this.items[this.first] = null
    this.first = (this.first + 1) % this.items.size
    this.size -= 1
    return item as T?
  }

  internal fun enqueue(item: T?): Unit {

    if (this.isFull) {
      throw Exception("Cannot enqueue into a full queue!")
    }

    this.last = (this.last + 1) % this.items.size
    this.items[this.last] = item
    this.size += 1
  }

}

File: Queue/queue.js
class Queue {

  constructor(length) {
    this._items  = new Array(length);
    this._last   = length - 1;
    this._first  = 0;
    this._length = 0;
  }

  get length() {
    return this._length;
  }

  get isEmpty() {
    return this._length === 0;
  }

  get isFull() {
    return this._length === this._items._length;
  }

  peek() {

    if (this.isEmpty) {
      throw new RangeError("Cannot peek at an empty queue!");
    }

    return this._items[this._first];
  }

  dequeue() {

    if (this.isEmpty) {
      throw new RangeError("Cannot dequeue from an empty queue!");
    }

    const item = this._items[this._first];
    this._items[this._first] = undefined;
    this._first = (this._first + 1) % this._items._length;
    this._length -= 1;
    return item;
  }

  enqueue(item) {

    if (this.isFull) {
      throw RangeError("Cannot enqueue into a full queue!");
    }

    this._last = (this._last + 1) % this._items._length;
    this._items[this._last] = item;
    this._length += 1;
  }

}
